massa-module-cache/src/hd_cache.rs

```diff
@@ -0,0 +1,301 @@
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+use rand::RngCore;
+use rocksdb::{Direction, IteratorMode, WriteBatch, DB};
+use std::path::PathBuf;
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const DATA_SER_ERROR: &str = "critical: metadata serialization failed";
+const DATA_DESER_ERROR: &str = "critical: metadata deserialization failed";
+const MOD_SER_ERROR: &str = "critical: module serialization failed";
+const MOD_DESER_ERROR: &str = "critical: module deserialization failed";
+const MODULE_IDENT: u8 = 0u8;
+const DATA_IDENT: u8 = 1u8;
+
+/// Module key formatting macro
+#[macro_export]
+macro_rules! module_key {
+    ($bc_hash:expr) => {
+        [&$bc_hash.to_bytes()[..], &[MODULE_IDENT]].concat()
+    };
+}
+
+/// Delta key formatting macro
+#[macro_export]
+macro_rules! metadata_key {
+    ($bc_hash:expr) => {
+        [&$bc_hash.to_bytes()[..], &[DATA_IDENT]].concat()
+    };
+}
+
+pub(crate) struct HDCache {
+    /// RocksDB database
+    db: DB,
+    /// How many entries are in the db. Count is initialized at creation time by iterating
+    /// over all the entries in the db then it is maintained in memory
+    entry_count: usize,
+    /// Maximum number of entries we want to keep in the db.
+    /// When this maximum is reached `snip_amount` entries are removed
+    max_entry_count: usize,
+    /// How many entries are removed when `entry_count` reaches `max_entry_count`
+    snip_amount: usize,
+    /// Module metadata serializer
+    meta_ser: ModuleMetadataSerializer,
+    /// Module metadata deserializer
+    meta_deser: ModuleMetadataDeserializer,
+}
+
+impl HDCache {
+    /// Create a new HDCache
+    ///
+    /// # Arguments
+    /// * path: where to store the db
+    /// * max_entry_count: maximum number of entries we want to keep in the db
+    /// * amount_to_remove: how many entries are removed when `entry_count` reaches `max_entry_count`
+    pub fn new(path: PathBuf, max_entry_count: usize, snip_amount: usize) -> Self {
+        let db = DB::open_default(path).expect(OPEN_ERROR);
+        let entry_count = db.iterator(IteratorMode::Start).count();
+
+        Self {
+            db,
+            entry_count,
+            max_entry_count,
+            snip_amount,
+            meta_ser: ModuleMetadataSerializer::new(),
+            meta_deser: ModuleMetadataDeserializer::new(),
+        }
+    }
+
+    /// Insert a new module in the cache
+    pub fn insert(&mut self, hash: Hash, module_info: ModuleInfo) {
+        if self.entry_count >= self.max_entry_count {
+            self.snip();
+        }
+
+        let mut ser_metadata = Vec::new();
+        let ser_module = match module_info {
+            ModuleInfo::Invalid => {
+                self.meta_ser
+                    .serialize(&ModuleMetadata::Invalid, &mut ser_metadata)
+                    .expect(DATA_SER_ERROR);
```

Comment:
Should we panic in case of an insert error in the cache? I would have thought it is not an unrecoverable error.

Reply:
The only causes of a module metadata serialization failure are:
* Internal programmer mistake
* Corrupted database

These errors were initially recovered but it was changed since they are critical and should never result in anything else than a panic.

Reply:
Ok, that makes sense!